File: week4/encapsulation-python/bank_accounts.py
class BankAccount:
    def __init__(self, owner, account_type, starting_balance=0):
        if starting_balance < 0:
            raise ValueError("starting_balance must be >= 0")
        self.owner = owner
        self.account_type = account_type
        self._balance = starting_balance
        self._history = ["Account opened"]

    @property
    def balance(self):
        return self._balance

    def deposit(self, amount):
        if amount <= 0:
            raise ValueError("amount must be > 0")
        self._balance += amount
        self._history.append(f"Deposited ${amount}")

    def withdraw(self, amount):
        if amount <= 0:
            raise ValueError("amount must be > 0")
        if amount > self._balance:
            raise ValueError("insufficient funds")
        self._balance -= amount
        self._history.append(f"Withdrew ${amount}")

    def apply_monthly_update(self):
        pass

    def history(self):
        return list(self._history)

    def __repr__(self):
        return f"BankAccount(owner={self.owner!r}, type={self.account_type}, balance={self._balance})"

    def __add__(self, other):
        if not isinstance(other, BankAccount):
            return NotImplemented
        return self._balance + other._balance


class TravelAccount(BankAccount):
    def __init__(self, owner, overdraft_limit, monthly_fee, starting_balance=0):
        super().__init__(owner, "Travel", starting_balance)
        self.overdraft_limit = overdraft_limit
        self.monthly_fee = monthly_fee

    def withdraw(self, amount):
        if amount <= 0:
            raise ValueError("amount must be > 0")
        if amount > self._balance + self.overdraft_limit:
            raise ValueError("overdraft limit exceeded")
        self._balance -= amount
        self._history.append(f"Withdrew ${amount}")

    def apply_monthly_update(self):
        if self.monthly_fee <= 0:
            return
        if self.monthly_fee > self._balance + self.overdraft_limit:
            return
        self._balance -= self.monthly_fee
        self._history.append(f"Monthly fee applied: ${self.monthly_fee}")


class RewardsAccount(BankAccount):
    def __init__(self, owner, bonus_rate, free_withdrawals, withdrawal_fee, starting_balance=0):
        super().__init__(owner, "Rewards", starting_balance)
        self.bonus_rate = bonus_rate
        self.free_withdrawals = free_withdrawals
        self.withdrawal_fee = withdrawal_fee
        self._withdrawals_this_month = 0

    def deposit(self, amount):
        super().deposit(amount)
        if self.bonus_rate > 0:
            bonus = amount * self.bonus_rate
            if bonus > 0:
                self._balance += bonus
                self._history.append(f"Rewards bonus applied: ${bonus}")

    def withdraw(self, amount):
        super().withdraw(amount)
        self._withdrawals_this_month += 1
        if self._withdrawals_this_month > self.free_withdrawals and self.withdrawal_fee > 0:
            if self.withdrawal_fee > self._balance:
                raise ValueError("insufficient funds for fee")
            self._balance -= self.withdrawal_fee
            self._history.append(f"Withdrawal fee applied: ${self.withdrawal_fee}")

    def apply_monthly_update(self):
        self._withdrawals_this_month = 0


class StudentAccount(BankAccount):
    def __init__(self, owner, monthly_interest_rate, daily_withdrawal_limit, starting_balance=0):
        super().__init__(owner, "Student", starting_balance)
        self.monthly_interest_rate = monthly_interest_rate
        self.daily_withdrawal_limit = daily_withdrawal_limit

    def withdraw(self, amount):
        if amount > self.daily_withdrawal_limit:
            raise ValueError("daily limit exceeded")
        super().withdraw(amount)

    def apply_monthly_update(self):
        if self.monthly_interest_rate <= 0:
            return
        interest = self._balance * self.monthly_interest_rate
        if interest > 0:
            self._balance += interest
            self._history.append(f"Monthly interest applied: ${interest}")


def run_demo():
    accounts = [
        TravelAccount("Avery", overdraft_limit=250, monthly_fee=12.5, starting_balance=500),
        RewardsAccount("Jordan", bonus_rate=0.02, free_withdrawals=2, withdrawal_fee=1.5, starting_balance=500),
        StudentAccount("Taylor", monthly_interest_rate=0.01, daily_withdrawal_limit=100, starting_balance=500),
    ]

    for account in accounts:
        account.deposit(75)
        account.withdraw(50)
        account.apply_monthly_update()
        print(account)

    combined = accounts[0] + accounts[1]
    print(f"Combined balance (Travel + Rewards): ${combined}")
    print("Operator overloading can be helpful, but clear methods are usually better in banking.")


if __name__ == "__main__":
    run_demo()

File: week4/encapsulation-javascript/bankAccounts.js
class BankAccount {
  constructor(owner, accountType, startingBalance = 0) {
    if (startingBalance < 0) {
      throw new Error("startingBalance must be >= 0");
    }
    this.owner = owner;
    this.accountType = accountType;
    this._balance = startingBalance;
    this._history = ["Account opened"];
  }

  get balance() {
    return this._balance;
  }

  deposit(amount) {
    if (amount <= 0) {
      throw new Error("amount must be > 0");
    }
    this._balance += amount;
    this._history.push(`Deposited $${amount}`);
  }

  withdraw(amount) {
    if (amount <= 0) {
      throw new Error("amount must be > 0");
    }
    if (amount > this._balance) {
      throw new Error("insufficient funds");
    }
    this._balance -= amount;
    this._history.push(`Withdrew $${amount}`);
  }

  applyMonthlyUpdate() {}

  history() {
    return [...this._history];
  }

  toString() {
    return `Account(${this.accountType}) - ${this.owner}: $${this._balance}`;
  }

  static combineBalances(left, right) {
    return left._balance + right._balance;
  }
}

class TravelAccount extends BankAccount {
  constructor(owner, overdraftLimit, monthlyFee, startingBalance = 0) {
    super(owner, "Travel", startingBalance);
    this.overdraftLimit = overdraftLimit;
    this.monthlyFee = monthlyFee;
  }

  withdraw(amount) {
    if (amount <= 0) {
      throw new Error("amount must be > 0");
    }
    if (amount > this._balance + this.overdraftLimit) {
      throw new Error("overdraft limit exceeded");
    }
    this._balance -= amount;
    this._history.push(`Withdrew $${amount}`);
  }

  applyMonthlyUpdate() {
    if (this.monthlyFee <= 0) {
      return;
    }
    if (this.monthlyFee > this._balance + this.overdraftLimit) {
      return;
    }
    this._balance -= this.monthlyFee;
    this._history.push(`Monthly fee applied: $${this.monthlyFee}`);
  }
}

class RewardsAccount extends BankAccount {
  constructor(owner, bonusRate, freeWithdrawals, withdrawalFee, startingBalance = 0) {
    super(owner, "Rewards", startingBalance);
    this.bonusRate = bonusRate;
    this.freeWithdrawals = freeWithdrawals;
    this.withdrawalFee = withdrawalFee;
    this._withdrawalsThisMonth = 0;
  }

  deposit(amount) {
    super.deposit(amount);
    if (this.bonusRate > 0) {
      const bonus = amount * this.bonusRate;
      if (bonus > 0) {
        this._balance += bonus;
        this._history.push(`Rewards bonus applied: $${bonus}`);
      }
    }
  }

  withdraw(amount) {
    super.withdraw(amount);
    this._withdrawalsThisMonth += 1;
    if (this._withdrawalsThisMonth > this.freeWithdrawals && this.withdrawalFee > 0) {
      if (this.withdrawalFee > this._balance) {
        throw new Error("insufficient funds for fee");
      }
      this._balance -= this.withdrawalFee;
      this._history.push(`Withdrawal fee applied: $${this.withdrawalFee}`);
    }
  }

  applyMonthlyUpdate() {
    this._withdrawalsThisMonth = 0;
  }
}

class StudentAccount extends BankAccount {
  constructor(owner, monthlyInterestRate, dailyWithdrawalLimit, startingBalance = 0) {
    super(owner, "Student", startingBalance);
    this.monthlyInterestRate = monthlyInterestRate;
    this.dailyWithdrawalLimit = dailyWithdrawalLimit;
  }

  withdraw(amount) {
    if (amount > this.dailyWithdrawalLimit) {
      throw new Error("daily limit exceeded");
    }
    super.withdraw(amount);
  }

  applyMonthlyUpdate() {
    if (this.monthlyInterestRate <= 0) {
      return;
    }
    const interest = this._balance * this.monthlyInterestRate;
    if (interest > 0) {
      this._balance += interest;
      this._history.push(`Monthly interest applied: $${interest}`);
    }
  }
}

function runDemo() {
  const accounts = [
    new TravelAccount("Avery", 250, 12.5, 500),
    new RewardsAccount("Jordan", 0.02, 2, 1.5, 500),
    new StudentAccount("Taylor", 0.01, 100, 500)
  ];

  for (const account of accounts) {
    account.deposit(75);
    account.withdraw(50);
    account.applyMonthlyUpdate();
    console.log(account.toString());
  }

  const combined = BankAccount.combineBalances(accounts[0], accounts[1]);
  console.log(`Combined balance (Travel + Rewards): $${combined}`);
  console.log("Operator overloading isn't in JS; use explicit methods for clarity.");
}

if (require.main === module) {
  runDemo();
}
